### PR TITLE
Reimplement network modifier `$to`

### LINKED ISSUE #4869: `$to` and `$denyallow` requires `$domain`
By the specification, `$domain` (or `$from`) is required to use `$to` and `$denyallow`. We're not aligning with it.

refs https://github.com/gorhill/ublock/wiki/static-filter-syntax#denyallow

### PR DESCRIPTION
**Pull request by seia-soto**
This addresses the regression rolled backed in #4828. Also this addresses #4869.

Since `$to` is a superset of `$denyallow`, I try to follow the baseline requirements from `$denyallow`. The important thing to validate with `$denyallow` is that it must be used with `$domain`. Still, it's possible to match filters without `$domain` requirement but the filter goes too wide.

In previous implementation, the values were transformed into a single `$denyallow` value by applying negation. If the domain entity was positive value, it was translated into `$denyallow` negative value and the negative went to positive value.

However, this violates the requirement of `$denyallow` that we need to ensure the existence of `$domain` and made the filter too wide to apply. Therefore, `$to` modifier was unintentionally applied to every hostname with `notHostnames` match in `Domains` data type.

The values and negated values in `$to` modifier is to address `requestDomains` and `excludedRequestDomains` field in declarative network request API: see https://github.com/gorhill/ublock/wiki/static-filter-syntax#to. `$denyallow` basically means "exceptions". Therefore, I translate `$to` values into `$domain` (or `$from`) values, and negated `$to` values into `$denyallow`.

fixes #4869

### CODE REVIEW

**Comment by seia-soto** [comment]
The problematic case is addressed here:

https://github.com/ghostery/adblocker/pull/4857/files#diff-a867a7bf156af04435bf054d829b5fe2f67e5caefff359ea39913d0149f699ccR902-R916

File: packages/adblocker/src/filters/network.ts [hunk 1]
```diff
@@ -727,27 +727,35 @@ export default class NetworkFilter implements IFilter {
       // --------------------------------------------------------------------- //
       // parseOptions
       // --------------------------------------------------------------------- //
+      const domainsList: Set<string> = new Set();
+      const denyallowList: Set<string> = new Set();
```
Comment (remusao) [inline packages/adblocker/src/filters/network.ts:731]: Nit. Can we create an empty set lazily so that it's not needed for the vast majority of filters? We could set them to `undefined` by default.

File: packages/adblocker/src/filters/network.ts [hunk 2]
```diff
@@ -966,6 +974,30 @@ export default class NetworkFilter implements IFilter {
           }
         }
       }
+
+      if (domainsList.size !== 0) {
+        domains = Domains.parse(Array.from(domainsList).join('|'), {
```
Comment (remusao) [inline packages/adblocker/src/filters/network.ts:979]: Would it make sense to change the API of `Domains.parse` to avoid having to re-join them but instead pass a `Set` or array?

File: packages/adblocker/test/matching.test.ts [hunk 3]
```diff
@@ -396,6 +396,23 @@ describe('#NetworkFilter.match', () => {
       url: 'https://sub.y.com/bar',
       type: 'script',
     });
+
+    // to
```
Comment (remusao) [inline packages/adblocker/test/matching.test.ts:400]: Should we add more tests? Not sure if we cover all corner cases with these 3 tests?

**Review by remusao — APPROVED** [review]

File: packages/adblocker/src/filters/network.ts (hunk 1 above)
Comment (seia-soto) [reply packages/adblocker/src/filters/network.ts]: updated in 025db7c5a9c0213b5cf228e1913d2aef9e17a0e1

File: packages/adblocker/src/filters/network.ts (hunk 2 above)
Comment (seia-soto) [reply packages/adblocker/src/filters/network.ts]: updated in 043710ab7a36208b4818e91b32ee8d313c225aba

File: packages/adblocker/src/engine/domains.ts [hunk 4]
```diff
@@ -13,15 +13,17 @@ import { binLookup, hasUnicode, HASH_INTERNAL_MULT } from '../utils.js';
 
 export class Domains {
   public static parse(
-    value: string,
+    value: string | Set<string>,
```
Comment (chrmod) [inline packages/adblocker/src/engine/domains.ts:16]: should it be `string[]` instead? one type, no string

Comment (seia-soto) [reply packages/adblocker/src/engine/domains.ts:16]: We don't pass `string[]` type. Alternatively, we can refactor them to accept `string[]` without a delim. However, we still need to pass a delim to the object as we need to care about `toString`.

**Review by chrmod — APPROVED** [review]
